src/Processors/Transforms/DistinctSortedTransform.cpp

```diff
@@ -0,0 +1,191 @@
+#include <Processors/Transforms/DistinctSortedTransform.h>
+#include <Columns/IColumn.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+extern const int SET_SIZE_LIMIT_EXCEEDED;
+}
+
+static bool less(const MutableColumns & lhs, const Columns & rhs, size_t i, size_t j, const SortDescription & descr)
+{
+    size_t key_ind = 0;
+    for (const auto & elem : descr)
+    {
+        size_t ind = elem.column_number;
+        int res = elem.direction * lhs[key_ind++]->compareAt(i, j, *rhs[ind], elem.nulls_direction);
+        if (res < 0)
+            return true;
+        else if (res > 0)
+            return false;
+    }
+    return false;
+}
+
+DistinctSortedTransform::DistinctSortedTransform(
+    const Block &header_,
+    const SizeLimits & set_size_limits_,
+    UInt64 limit_hint_,
+    const SortDescription & sorted_columns_descr_,
+    const Names & source_columns)
+    : ISimpleTransform(header_, header_, true)
+    , limit_hint(limit_hint_)
+    , set_size_limits(set_size_limits_)
+    , sorted_columns_descr(sorted_columns_descr_)
+{
+    /// Calculate sorted columns positions
+    sorted_columns_pos.reserve(sorted_columns_descr.size());
+    for (auto & sorted_column_descr : sorted_columns_descr)
+    {
+        if (!sorted_column_descr.column_name.empty())
+        {
+            sorted_column_descr.column_number = header_.getPositionByName(sorted_column_descr.column_name);
+            sorted_column_descr.column_name.clear();
+        }
+        sorted_columns_pos.emplace_back(sorted_column_descr.column_number);
+    }
+
+    /// Calculate not-sorted columns positions
+    other_columns_pos.reserve(source_columns.size());
+    for (const auto & source_column : source_columns)
+    {
+        size_t pos = header_.getPositionByName(source_column);
+        if (std::find(sorted_columns_pos.begin(), sorted_columns_pos.end(), pos) != sorted_columns_pos.end())
+            continue;
+
+        const auto & col = header_.getByPosition(pos).column;
+        if (!(col && isColumnConst(*col)))
```

Comment:
What if `col` will be `nullptr`.  It's impossible, but technically condition is not correct. Looks like it shoud be `col && !isColumnConst(*col).